### PR TITLE
[fga] WorkspaceService.controlAdmission

### PR DESCRIPTION
**Pull request by geropl**
## Description

This introduced WorkspaceService.controlAdmission, incl. workspace sharing implemented with FGA

<details>
<summary>Summary generated by Copilot</summary>

### <samp>🤖 Generated by Copilot at 73ce811</samp>

This pull request adds a new feature to allow sharing workspaces with any user in an organization. It modifies the authorization logic and database schema for workspaces, and adds new methods and tests to the `WorkspaceService` and `Authorizer` classes. It also updates the Spicedb schema and permissions to support the new `shared` relation.

</details>

## Related Issue(s)
Relates to: EXP-207

## How to test
 - create user's 1 and 2
 - user 1: start a workspace, and enable sharing :heavy_check_mark: 
 - user 2: open the share link, and see how you can properly interact with the workspace :heavy_check_mark: 
 - user 1: disable workspace sharing :heavy_check_mark: 
 - user 2: re-load the workspace page, and note how you lost access :heavy_check_mark: 
 
#### Preview status

<p>Gitpod was successfully deployed to your preview environment.</p>
<ul>
	<li><b>🏷️ Name</b> - gpl-fga-share-ws</li>
	<li><b>🔗 URL</b> - <a href="https://gpl-fga-share-ws.preview.gitpod-dev.com/workspaces" target="_blank">gpl-fga-share-ws.preview.gitpod-dev.com/workspaces</a>.</li>
	<li><b>📚 Documentation</b> - See our <a href="https://www.notion.so/gitpod/6debd359591b43688b52f76329d04010#7c1ce80ab31a41e29eff2735e38eec39" target="_blank">internal documentation</a> for information on how to interact with your preview environment.</li>
	<li><b>📦 Version</b> - gpl-fga-share-ws-gha.15657</li>
</ul>

## Build Options

<details>
<summary>Build</summary>

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`
</details>

<details>
<summary>Publish</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer</summary>

- [ ] analytics=segment
- [ ] with-dedicated-emulation
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

<details>
<summary>Preview Environment / Integration Tests</summary>

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [x] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`. If enabled, `with-preview` and `with-large-vm` will be enabled.
- [ ] with-monitoring
</details>

/hold

### CODE REVIEW

File: components/server/src/workspace/workspace-service.ts [hunk 1]
```diff
@@ -813,6 +819,53 @@ export class WorkspaceService {
...
         }
     }
+
+    public async controlAdmission(
+        userId: string,
+        workspaceId: string,
+        level: "owner" | "everyone",
+        check: (workspace: Workspace, instance?: WorkspaceInstance) => Promise<void> = async () => {},
+    ): Promise<void> {
+        await this.auth.checkPermissionOnWorkspace(userId, "access", workspaceId);
```
Comment (svenefftinge) [inline components/server/src/workspace/workspace-service.ts:829]: We should probably have a specific permission for changing the sharing properties in another PR. At least if we decide to keep this feature.

**Review by svenefftinge — APPROVED** [review]
Sweet! ❤️

File: components/server/src/workspace/workspace-service.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/workspace/workspace-service.ts:829]: Jep. Was on the brink to introduce one, then kept the current behavior. :+1:

**Pull request merged**